Expand `PaymentClaimable` to include all inbound channel IDs for a payment

shaavan opened the pull request:
Resolves #2274  

This PR expands `PaymentClaimable` to include all inbound `channel_ids` and `user_channel_ids` associated with a payment.  
This update ensures a clearer representation of the channels used, especially for MPP payments.

shaavan commented:
@benthecarman 

> Would also be nice if the same information is available on the PaymentClaimed event too

In the current version, I haven’t included `channel_ids` and `user_channel_ids` for `PaymentClaimed`, as they can be easily derived from the event itself:

```rust
let channel_ids = payment_claimed_event.htlcs.iter().map(|htlc| htlc.channel_id).collect();
```

That said, let me know if you’d still prefer to include `channel_ids` in `PaymentClaimed`, and I'd be happy to make that change. Thanks!

benthecarman commented:
Thanks, @shaavan , that is good enough

valentinewallace commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -2768,8 +2768,8 @@ pub fn do_pass_along_path<'a, 'b, 'c>(args: PassAlongPathArgs) -> Option<Event>
 						assert_eq!(*amount_msat, recv_value);
-						assert!(node.node.list_channels().iter().any(|details| details.channel_id == via_channel_id.unwrap()));
-						assert!(node.node.list_channels().iter().any(|details| details.user_channel_id == via_user_channel_id.unwrap()));
+						assert!(node.node.list_channels().iter().any(|details| details.channel_id == via_channel_ids.last().unwrap().unwrap()));
+						assert!(node.node.list_channels().iter().any(|details| details.user_channel_id == via_user_channel_ids.last().unwrap().unwrap()));
```
For completeness, could we iterate through the event channel ids list and make sure all of them are present in `list_channels`?

shaavan replied:
Updated in [pr3655.02](https://github.com/shaavan/rust-lightning/commits/pr3655.02)
Thanks for the pointer!

valentinewallace commented on lightning/src/events/mod.rs:
```diff
@@ -774,10 +774,16 @@ pub enum Event {
+		/// - In a non-MPP scenario, this will contain a single `channel_id` where the payment was received.
+		/// - In an MPP scenario, this will contain multiple `channel_id`s corresponding to the channels over
+		///   which the payment parts were received.
+		via_channel_ids: Vec<Option<ChannelId>>,
```
I think it might be a better API to do `Vec<ChannelId>` (similar below) and document for what versions the list will be incomplete or empty. To confirm, it will be a complete list of channels ids for events created on the version this code is released in onward?

shaavan replied:
I have updated the new fields to be `Vec<ChannelId>` and `Vec<u128>` in [pr3655.02](https://github.com/shaavan/rust-lightning/commits/pr3655.02).  
Thanks a lot for the pointer!

> To confirm, it will be a complete list of channel IDs for events created on the version this code is released in onward?

It looks like these fields were initially introduced as `Option<type>` in [#1856](https://github.com/lightningdevkit/rust-lightning/pull/1856) to maintain backward compatibility with older versions (see [this commit](https://github.com/lightningdevkit/rust-lightning/pull/1856/commits/0edb0e2f8442e1e1267f19c99c705faa1ce36930)). So, as far as I understand, the list will be complete for all cases where HTLCs are created using version `0.0.113` and later.

Let me know if I’ve understood this correctly—I'd be happy to add a note for these fields. Thanks!

valentinewallace replied:
Yeah, that sounds right to me!

shaavan replied:
Done! Updated the docs in [pr3655.03](https://github.com/shaavan/rust-lightning/commits/pr3655.03)
Thanks a lot for the suggestion!

valentinewallace commented:
nit: limit commit subject line in 2nd commit to ~50 characters

shaavan commented:
**Updated from** [pr3655.01](https://github.com/shaavan/rust-lightning/commits/pr3655.01) to [pr3655.02](https://github.com/shaavan/rust-lightning/commits/pr3655.02) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3655.01..pr3655.02)):
Addressed @valentinewallace comments

1. Updated the fields to be `Vec<ChannelId>`, and `Vec<u128>`
2. Updated test utilities to check for all the channels id in events list, for improved test completeness.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -926,13 +926,13 @@ impl ClaimablePayment {
-	fn get_user_channel_ids(&self) -> Vec<Option<u128>> {
-		self.htlcs.iter().map(|htlc| htlc.prev_hop.user_channel_id).collect()
+	fn get_user_channel_ids(&self) -> Vec<u128> {
+		self.htlcs.iter().map(|htlc| htlc.prev_hop.user_channel_id.unwrap_or(0)).collect()
```
Would be nice to `filter_map` instead of falling back to inserting `0`, could also document on the method(s) for what versions the lists will be inaccurate, if any.

shaavan replied:
That looks cleaner! Updated in [pr3655.03](https://github.com/shaavan/rust-lightning/commits/pr3655.03)
Also, I have added a small note that the list will be incomplete for HTLC created using LDK versions `<= 0.0.112`

Thanks a lot for the suggestion!

valentinewallace commented on lightning/src/events/mod.rs:
```diff
@@ -1544,16 +1550,20 @@ impl Writeable for Event {
 					(2, payment_secret, option),
-					(3, via_channel_id, option),
+					// legacy field
+					// (3, via_channel_id, option),
 					(4, amount_msat, required),
-					(5, via_user_channel_id, option),
+					// legacy field
+					// (5, via_user_channel_id, option),
```
I think we should still write these fields for the sake of users who downgrade (can always write the first item in the vec). Also would be nice to comment in what version the fields became legacy.

valentinewallace commented on lightning/src/events/mod.rs:
```diff
@@ -774,10 +774,16 @@ pub enum Event {
-		/// The `user_channel_id` indicating over which channel we received the payment.
-		via_user_channel_id: Option<u128>,
+		/// The `channel_id`(s) indicating over which channel(s) we received the payment.
+		/// - In a non-MPP scenario, this will contain a single `channel_id` where the payment was received.
+		/// - In an MPP scenario, this will contain multiple `channel_id`s corresponding to the channels over
+		///   which the payment parts were received.
```
IMO we can delete this, similar below. Could we document for which versions these vecs will be empty as well, if any?

shaavan replied:
Done! [pr3655.03](https://github.com/shaavan/rust-lightning/commits/pr3655.03)
I have deleted the docs section and added a note that this will be empty vector for LDK version `<= 0.0.112`
Thanks for the pointer!

valentinewallace commented on lightning/src/events/mod.rs:
```diff
@@ -1855,6 +1865,8 @@ impl MaybeReadable for Event {
 					let mut onion_fields = None;
```
Maybe it's already obvious but for clarity we could suffix the old fields with `_legacy`.

shaavan replied:
Yep, that sounds cleaner! Done [pr3655.03](https://github.com/shaavan/rust-lightning/commits/pr3655.03)
Thanks! ✨

shaavan commented:
**Updated from** [pr3655.02](https://github.com/shaavan/rust-lightning/commits/pr3655.02) to [pr3655.03](https://github.com/shaavan/rust-lightning/commits/pr3655.03) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3655.02..pr3655.03)):
Addressed @valentinewallace comments

Changes:
1. Expanded Documentation
2. Properly write the legacy fields to maintain downgrade support.

valentinewallace commented:
Looks good, feel free to squash! Thanks!

shaavan commented:
**Updated from** [pr3655.03](https://github.com/shaavan/rust-lightning/commits/pr3655.03) to [pr3655.04](https://github.com/shaavan/rust-lightning/commits/pr3655.04) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3655.03..pr3655.04)):
Addressed @valentinewallace comments

Changes:
1. Squash commits.

shaavan commented on lightning/src/events/mod.rs:
```diff
@@ -1540,20 +1542,25 @@ impl Writeable for Event {
 					(1, receiver_node_id, option),
 					(2, payment_secret, option),
-					(3, via_channel_id, option),
+					(3, via_channel_id_legacy, option),
```
Quick question — are we going with `0.1.3` or `0.2.0` for the next update?  
I'll just add a quick comment here to note in which version this became `legacy`.
Thanks! ✨

valentinewallace replied:
Sorry for the delay, that'd be 0.2 https://discord.com/channels/915026692102316113/915026887066132481/1353786722705477713

shaavan replied:
Thanks for the pointer, Val! Added the comment in [pr3655.05](https://github.com/shaavan/rust-lightning/commits/pr3655.05) 🚀

shaavan reviewed (commented):
> Looks good, feel free to squash! Thanks!

Awesome — thanks a lot for the review, Val! 🚀

shaavan commented:
**Updated from** [pr3655.04](https://github.com/shaavan/rust-lightning/commits/pr3655.04) to [pr3655.05](https://github.com/shaavan/rust-lightning/commits/pr3655.05) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3655.04..pr3655.05)):
Addressed @valentinewallace comments

Changes:
- Added comment documenting when, and why `via_channel_id_legacy`, and `via_user_channel_id_legacy` were deprecated.

shaavan commented:
**Updated from** [pr3655.05](https://github.com/shaavan/rust-lightning/commits/pr3655.05) to [pr3655.06](https://github.com/shaavan/rust-lightning/commits/pr3655.06) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3655.05..pr3655.06)):
Addressed @valentinewallace comments

Changes:
1. Updated documentation to correctly reflect that the vector will be incomplete for HTLC created using versions prior to `LDK 0.1.0 or prior.`

shaavan commented on lightning/src/events/mod.rs:
```diff
@@ -774,10 +774,12 @@ pub enum Event {
+		/// This will be an empty vector for events created/serialised using LDK version 0.0.112 and prior.
+		via_channel_ids: Vec<ChannelId>,
+		/// The `user_channel_id`(s) corresponding to the channels over which the payment was received.
+		/// This will be an empty vector for HTLC events created/serialised using LDK version 0.0.112 and prior.
```
Makes sense! Thanks for the pointer!
Updated the docs in [pr3655.06](https://github.com/shaavan/rust-lightning/commits/pr3655.06)

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -924,6 +924,18 @@ impl ClaimablePayment {
+
+	/// Returns the inbound `user_channel_id`s for all HTLCs associated with the payment.
+	///
+	/// Note: This list will be incomplete for HTLCs created using LDK version 0.0.112 or earlier.
```
Done! [pr3655.06](https://github.com/shaavan/rust-lightning/commits/pr3655.06)
Thanks a lot for the pointer!

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -924,6 +924,20 @@ impl ClaimablePayment {
+
+	/// Returns the inbound `user_channel_id`s for all HTLCs associated with the payment.
+	///
+	/// Note: This list will be incomplete for HTLCs created using LDK version 0.1.0 or prior.
```
I looked through the git blame and it looks like the `user_channel_id` field was actually added to `HTLCPreviousHopData` in 117: https://github.com/lightningdevkit/rust-lightning/pull/2478/commits/2e562a2ad88fe214e19e2ee1c5f4db743d9ebe07 so this list will be incomplete for HTLCs created prior to 117 instead.

shaavan replied:
Makes sense! I initially assumed the list would be incomplete for all versions before we added the `get_...` fetcher, but you're right — since the data has been available within the HTLC itself starting from `0.0.117`, the list should be complete for HTLCs created post-`0.0.117`.  
Thanks for the pointer! Corrected in [pr3655.07](https://github.com/shaavan/rust-lightning/commits/pr3655.07).

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -924,6 +924,20 @@ impl ClaimablePayment {
 		)
 	}
+
+	/// Returns the inbound `channel_id`s for all HTLCs associated with the payment.
+	///
+	/// Note: This list will be incomplete for HTLCs created using LDK version 0.1.0 or prior.
+	fn get_channel_ids(&self) -> Vec<ChannelId> {
+		self.htlcs.iter().map(|htlc| htlc.prev_hop.channel_id).collect()
+	}
```
This list should actually always be complete, because it looks like the `HTLCPreviousHopData::channel_id` will always be set as long as the struct's `outpoint` field is set, and that field has existed since 2020. So we can remove this note. Sorry for the back and forth here!

shaavan replied:
Ah, that makes sense — thanks for clarifying! I hadn’t traced it back that far, so this definitely helps.  
No worries at all about the back and forth — I’m grateful for the eagle-eyed catch! I’ve gone ahead and removed the note in [pr3655.07](https://github.com/shaavan/rust-lightning/commits/pr3655.07).

shaavan commented:
**Updated from** [pr3655.06](https://github.com/shaavan/rust-lightning/commits/pr3655.06) to [pr3655.07](https://github.com/shaavan/rust-lightning/commits/pr3655.07) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3655.06..pr3655.07)):
Addressed @valentinewallace comments

Changes:
1. Fixed comments to correctly represent for which version and cases the lists would be incomplete.

shaavan commented on lightning/src/events/mod.rs:
```diff
@@ -774,10 +774,12 @@ pub enum Event {
-		/// The `user_channel_id` indicating over which channel we received the payment.
-		via_user_channel_id: Option<u128>,
+		/// The `channel_id`(s) over which the payment was received.
+		/// This will be an incomplete vector for events created/serialized using LDK version 0.1.0 and prior.
```
Got it! Updated in [pr3655.07](https://github.com/shaavan/rust-lightning/commits/pr3655.07)
Thanks!

joostjager commented on lightning/src/events/mod.rs:
```diff
@@ -1540,20 +1542,29 @@ impl Writeable for Event {
 					(11, payment_context, option),
 					(13, payment_id, option),
+					(15, *via_channel_ids, optional_vec),
+					(17, *via_user_channel_ids, optional_vec),
```
When I comment out these two new lines, no tests seem to fail. Probably no coverage of this change?

shaavan replied:
Thanks a lot for the pointer!
Updated the test-suite coverage, so that we check for proper event serialisation-deserialisation in [pr3655.08](https://github.com/shaavan/rust-lightning/commits/pr3655.08)

joostjager commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -2068,11 +2068,11 @@ pub fn test_channel_reserve_holding_cell_htlcs() {
 			assert_eq!(recv_value_21, amount_msat);
 			assert_eq!(nodes[2].node.get_our_node_id(), receiver_node_id.unwrap());
-			assert_eq!(via_channel_id, Some(chan_2.2));
+			assert_eq!(*via_channel_ids.last().unwrap(), chan_2.2);
```
It seems there is no coverage for via_channel_ids in an actual mpp scenario. Perhaps there is a test to which this can be added in a simple way?

shaavan replied:
Hi Joost!
Thanks for pointing this out — I just took a look through the codebase, and it seems we do actually check for MPP scenarios involving the new `via_channel_ids` and `via_user_channel_ids`.

For instance, in the test [`mpp_to_one_hop_blinded_path`](https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/ln/blinded_payment_tests.rs#L196), after the second `pass_along_path()` call, the `PaymentClaimable` event that's generated includes a `via_channel_ids` field with a `Vec` of two channel IDs.

That said, if you had something else in mind or if there's a better place we could be testing this, I’m happy to dig further!

joostjager replied:
IIRC correctly what I did while reviewing initially is that I changed the code to only set `via_channel_ids` to the first channel, and then didn't see any tests fail.

joostjager reviewed (commented):
Code looks good, test coverage could probably be improved a bit.

shaavan commented:
**Updated from** [pr3655.07](https://github.com/shaavan/rust-lightning/commits/pr3655.07) to [pr3655.08](https://github.com/shaavan/rust-lightning/commits/pr3655.08) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3655.07..pr3655.08)):
Addressed @joostjager comments

Changes:
1. Expanded event test-suite, to check that all events are correctly serialised-deserialised.

shaavan commented:
**Updated from** [pr3655.08](https://github.com/shaavan/rust-lightning/commits/pr3655.08) to [pr3655.09](https://github.com/shaavan/rust-lightning/commits/pr3655.09) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3655.08..pr3655.09)):
Addressed @joostjager comments

Changes:
1. Expanded test, to ensure that all the via_channel_id & via_user_channel_ids are present in PaymentClaimable in multi part payment case.

valentinewallace commented:
LGTM

joostjager commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10929,7 +10929,18 @@ where
+		// we test all generated events round-trip:
+		for event in &collected_events {
+			let ser = event.encode();
+			if let Some(deser) = events::Event::read(&mut &ser[..]).expect("event should deserialize") {
```
Is the None case ok to ignore?

valentinewallace replied:
Yeah, some events are never written so I believe @shaavan is correct here

joostjager replied:
So could also continue when `ser` is None then, and unwrap the read? Not that it is all that important ofc.

valentinewallace replied:
`ser` just returns `Ok(())` on success, not an option. Also we do currently unwrap the read with the `expect`.. But if you want to rearrange this code somehow, no strong preference here.

joostjager replied:
It looks like `ser` is just `Vec<u8>`. So if that is empty, we expect `None` from `read`? Maybe that can be used to make it slightly stricter and detect the case where read is unexpectedly returning None.

valentinewallace replied:
Oh right, sorry. It looks like for events where we don't write the actual event, we will still write the type for the event: https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/events/mod.rs#L1809 so I don't think we can check if it's empty but we could check if only a u8 was written.

joostjager replied:
Perhaps a too much internal detail then. Can leave this as is as far as I am concerned.

shaavan commented:
**Updated from** [pr3655.09](https://github.com/shaavan/rust-lightning/commits/pr3655.09) to [pr3655.10](https://github.com/shaavan/rust-lightning/commits/pr3655.10) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3655.09..pr3655.10)):
Addressed @joostjager comments

Changes:
1. Add check that the `via_channel_ids`, and `via_user_channel_ids` only contains one element for the appropriate cases.

shaavan commented on lightning/src/ln/chanmon_update_fail_tests.rs:
```diff
@@ -547,11 +547,11 @@ fn do_test_monitor_temporary_update_fail(disconnect_count: usize) {
 			assert_eq!(amount_msat, 1_000_000);
 			assert_eq!(receiver_node_id.unwrap(), nodes[1].node.get_our_node_id());
-			assert_eq!(via_channel_id, Some(channel_id));
+			assert_eq!(*via_channel_ids.last().unwrap(), channel_id);
```
Makes sense! Addressed in [pr3655.10](https://github.com/shaavan/rust-lightning/commits/pr3655.10)
Thanks!

joostjager commented on lightning/src/events/mod.rs:
```diff
@@ -774,10 +774,12 @@ pub enum Event {
+		via_channel_ids: Vec<ChannelId>,
+		/// The `user_channel_id`(s) corresponding to the channels over which the payment was received.
+		/// This will be an incomplete vector for MPP payment events created/serialized using LDK version 0.1.0 and prior.
+		via_user_channel_ids: Vec<u128>,
```
The option would be consistent with `HTLCPreviousHopData` for example

joostjager replied:
Also, isn't it the case that pre-117 data just translates to an empty vec<(..,..)> without requiring the option?

joostjager commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -924,6 +924,18 @@ impl ClaimablePayment {
+
+	/// Returns the inbound `user_channel_id`s for all HTLCs associated with the payment.
+	///
+	/// Note: This list will be incomplete for HTLCs created using LDK version 0.0.117 or prior.
```
This comment does seem to suggest that vecs can be different lengths?

joostjager replied:
@shaavan, @valentinewallace and I chatted about it. Seems something to say for a stronger type, especially because the vecs both map from `PrevHopData` already. But it's a bit late now also, and not too critical. So leave up to you to decide.

shaavan replied:
Thanks so much, Joost and Val, for the pointers!
I went ahead and updated the `PaymentClaimable` field to a `via_channel_id_pairs`, containing a `Vec<(ChannelId, Option<u128>)>`, in [pr3655.11](https://github.com/shaavan/rust-lightning/commits/pr3655.11).

Let me know if it looks good — thanks again!

shaavan commented:
**Updated from** [pr3655.10](https://github.com/shaavan/rust-lightning/commits/pr3655.10) to [pr3655.11](https://github.com/shaavan/rust-lightning/commits/pr3655.11) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3655.10..pr3655.11)):
Addressed @valentinewallace, @joostjager comments

Changes:
1. Update PaymentClaimable field to be an ordered `via_channel_id_pairs` pair.

joostjager commented:
@shaavan just a side comment: I really like how you offer diffs to reviewers using pr3655.x branches!

joostjager commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -924,6 +924,15 @@ impl ClaimablePayment {
+	/// Returns the inbound `(channel_id, user_channel_id)` pairs for all HTLCs associated with the payment.
+	///
+	/// Note: The `user_channel_id` will be `None` for HTLCs created using LDK version 0.0.117 or prior.
+	fn get_channel_id_pairs(&self) -> Vec<(ChannelId, Option<u128>)> {
```
Thinking about naming. `pairs` seems a bit redundant because the type is also a pair. But don't know anything better. Or just stick with ids?

Not a rust expert, but I believe that `get_` isn't a convention in rust? Could name this (in line with `inbound_payment_id` above) `inbound_channel_ids`?

shaavan replied:
Makes sense! I updated the names to `inbound_channel_ids` in [pr3655.12](https://github.com/shaavan/rust-lightning/commits/pr3655.12)
Thanks a lot for the pointer!

joostjager commented on lightning/src/events/mod.rs:
```diff
@@ -774,10 +774,10 @@ pub enum Event {
+		/// The `(channel_id, user_channel_id)` pairs over which the payment was received.
+		///
+		/// This will be an incomplete vector for MPP payment events created/serialized using LDK version 0.1.0 and prior.
+		via_channel_id_pairs: Vec<(ChannelId, Option<u128>)>,
```
Also standardize on `inbound_*` perhaps. Via is more broad. Although pre-existing also.

shaavan replied:
Thanks for the suggestion!
Updated the name to `inbound_channel_ids` in [pr3655.12](https://github.com/shaavan/rust-lightning/commits/pr3655.12)

joostjager commented on lightning/src/ln/chanmon_update_fail_tests.rs:
```diff
@@ -165,11 +165,12 @@ fn do_test_simple_monitor_temporary_update_fail(disconnect: bool) {
 			assert_eq!(receiver_node_id.unwrap(), nodes[1].node.get_our_node_id());
-			assert_eq!(via_channel_id, Some(channel_id));
+			assert_eq!(via_channel_id_pairs.len(), 1);
+			assert_eq!(via_channel_id_pairs.last().unwrap().0, channel_id);
```
Can't this simply be `assert_eq!(via_channel_id_pairs, vec![(channel_id, other_value)]);` ?

shaavan replied:
That makes sense! Also this help us expand the test suite, covering for user_channel_id, also with channel_id
Updated in [pr3655.12](https://github.com/shaavan/rust-lightning/commits/pr3655.12). Thanks for the pointer!

joostjager commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -240,8 +240,28 @@ fn mpp_to_one_hop_blinded_path() {
+				.collect();
+
+			assert_eq!(actual_chan_ids, expected_chan_ids);
+			assert_eq!(actual_user_chan_ids, expected_user_chan_ids);
```
Isn't it possible to do a vec<tuple> comparison here to avoid splitting everything into two lists?

shaavan replied:
Thanks for the pointer, Joost!

Since `list_channels` returns channels in arbitrary order, I wanted the comparison to be order-agnostic.  
In [pr3655.12](https://github.com/shaavan/rust-lightning/commits/pr3655.12), I updated the code to compare sorted vectors of the expected and actual `inbound_channel_ids` to achieve that. It makes the test cleaner and easier to reason about.

joostjager reviewed (approved):
Seeing how this worked out, I think vec<tuple> is the right move. Hopefully you feel the same!

joostjager reviewed (commented):
Accidentally approved. Wanted to await your replies first.

shaavan commented:
**Updated from** [pr3655.11](https://github.com/shaavan/rust-lightning/commits/pr3655.11) to [pr3655.12](https://github.com/shaavan/rust-lightning/commits/pr3655.12) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3655.11..pr3655.12)):
Addressed @joostjager comments

Changes:
1. Rename `via_channel_id_pairs` & `get_channel_id_pairs` -> inbound_channel_ids
2. Cleaned up testing.

shaavan commented:
> @shaavan just a side comment: I really like how you offer diffs to reviewers using pr3655.x branches!

Thanks so much, @joostjager — really appreciate you noticing that! :)

joostjager commented on lightning/src/ln/chanmon_update_fail_tests.rs:
```diff
@@ -547,11 +549,12 @@ fn do_test_monitor_temporary_update_fail(disconnect_count: usize) {
 			assert_eq!(amount_msat, 1_000_000);
 			assert_eq!(receiver_node_id.unwrap(), nodes[1].node.get_our_node_id());
-			assert_eq!(via_channel_id, Some(channel_id));
+			assert_eq!(inbound_channel_ids.len(), 1);
```
Redundant now?

shaavan replied:
Thanks for catching it! ✨
Looks like I forgot to remove this, when I update the code.
Fixed in [pr3655.13](https://github.com/shaavan/rust-lightning/commits/pr3655.13)

joostjager commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -2768,8 +2768,12 @@ pub fn do_pass_along_path<'a, 'b, 'c>(args: PassAlongPathArgs) -> Option<Event>
-						assert!(node.node.list_channels().iter().any(|details| details.user_channel_id == via_user_channel_id.unwrap()));
+						for (chan_id, user_chan_id) in inbound_channel_ids {
+							assert!(node.node.list_channels().iter().any(|details| &details.channel_id == chan_id));
+							if let Some(user_id) = user_chan_id {
```
Could probably first `node.node.list_channels().iter().find(|details| &details.channel_id == chan_id)` and then see of user_id matches too?

shaavan replied:
Makes sense! Updated in [pr3655.13](https://github.com/shaavan/rust-lightning/commits/pr3655.13)
Thanks!

joostjager commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -2068,11 +2068,11 @@ pub fn test_channel_reserve_holding_cell_htlcs() {
 			assert_eq!(recv_value_21, amount_msat);
 			assert_eq!(nodes[2].node.get_our_node_id(), receiver_node_id.unwrap());
-			assert_eq!(via_channel_id, Some(chan_2.2));
+			assert_eq!(inbound_channel_ids.last().unwrap().0, chan_2.2);
```
The reason you use `last()` is because there are multiple elements in the vec?

shaavan replied:
I initially used `last` here because, before this PR, `via_channel_id` referred to the channel over which the *last* HTLC of the MPP was received.
But in [pr3655.13](https://github.com/shaavan/rust-lightning/commits/pr3655.13), I updated it to use the vector comparison approach — consistent with the rest of the PR.

Thanks a lot for the pointer!

joostjager reviewed (commented):
LGTM besides a few nits.

shaavan commented:
**Updated from** [pr3655.12](https://github.com/shaavan/rust-lightning/commits/pr3655.12) to [pr3655.13](https://github.com/shaavan/rust-lightning/commits/pr3655.13) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3655.12..pr3655.13)):
Addressed @joostjager comments

Changes:
1. Cleaned up tests

valentinewallace reviewed (approved):
Thanks @shaavan! I like the tuple change 👌

joostjager reviewed (approved):
Thank you for this contribution. Great work and excellent communication.
